### PR TITLE
Fix everyone count for subadmins

### PR DESCRIPTION
**Pull request by PVince81**
Also moved the logic to the UsersController

@LukasReschke

### CODE REVIEW

**Comment by PVince81** [comment]
I'm getting a 403 for the call from the UI. I also tried using another URL like "/settings/userstats" to avoid conflicting with "/settings/users/*" but same result.

**Comment by PVince81** [comment]
Aha, it works for the regular admin but not subadmin. Will double check the directives.

**Comment by PVince81** [comment]
@LukasReschke I don't understand: the `function index()` only has `NoAdminRequired` and is called properly for both admin and subadmins.
But my new `function stats()` also has `NoAdminRequired` but for some reason subadmins are denied.

**Comment by PVince81** [comment]
Never mind, it's not a routing issue. It does call my method but am getting 403 inside, maybe accessing the wrong groups.

**Comment by PVince81** [comment]
Fixed and rebased onto master.

This PR does two things:
1) Moves the counting of all users to the UsersControllers, which makes it unit testable
2) Fix the counting of users from the subadmin's perspective: now users are unique even when inside multiple groups managed by the same subadmin

Fixes https://github.com/owncloud/core/issues/16566

Please review @LukasReschke @MorrisJobke @blizzz

File: settings/controller/userscontroller.php [hunk 1]
```diff
@@ -548,4 +548,40 @@ public function setMailAddress($id, $mailAddress) {
...
+		} else {
+			$groups = $this->groupManager->getSubAdmin()->getSubAdminsGroups($this->userSession->getUser());
+
+			foreach ($groups as $group) {
+				foreach($group->getUsers() as $uid => $displayName) {
+					$uniqueUsers[$uid] = true;
+				}
+			}
+
+			$userCount = count($uniqueUsers);
```
Comment (MorrisJobke) [inline settings/controller/userscontroller.php]: What happens if this is called by a normal user without admin/subadmin rights?

Comment (LukasReschke) [inline settings/controller/userscontroller.php]: Can't happen. There is a middleware in place that checks if the user is at least SubAdmin and a custom `@NoSubadminRequired` :)

Comment (MorrisJobke) [inline settings/controller/userscontroller.php]: Okay :) So an admin is automatically an subadmin?

**Comment by PVince81** [comment]
@LukasReschke @MorrisJobke @blizzz please review (original ticket was for 8.2.1)

File: settings/controller/userscontroller.php (hunk 1 above)
Comment (LukasReschke) [inline settings/controller/userscontroller.php]: Correct :)

https://github.com/owncloud/core/blob/c6f6a8758b3f08e47c3a8f45a67d09698376b2e2/lib/private/subadmin.php#L205-L208

**Comment by LukasReschke** [comment]
:+1:

**Comment by MorrisJobke** [comment]
I just fixed the case if an subadmin with no users in its groups browses the user management:

```
{"reqId":"mVhJYiVGpmG7vCBTbz9T","remoteAddr":"::1","app":"PHP","message":"Undefined variable: uniqueUsers at \/Users\/morrisjobke\/Projects\/owncloud\/master\/settings\/controller\/userscontroller.php#577","level":0,"time":"2015-11-04T08:38:01+00:00","method":"GET","url":"\/master\/index.php\/settings\/users\/stats"}
```

**Comment by MorrisJobke** [comment]
Works and fixes the counting problem :+1:

**Comment by PVince81** [comment]
@MorrisJobke ah yes, thanks for the fix. :+1: for that one too.

**Comment by PVince81** [comment]
Original ticket was set to 8.2.1, should I backport this ? @karlitschek

**Comment by karlitschek** [comment]
yes. please backport :+1:

**Comment by PVince81** [comment]
stable8.2: https://github.com/owncloud/core/pull/20328

**Pull request merged**